ui/app/src/services/configuration.ts

```diff
@@ -205,17 +205,28 @@ export function fetchSiteUiConfig(site: string): Observable<Pick<GlobalState['ui
           },
           launcher: null,
           dashboard: null,
-          datasets: {}
+          references: {}
         };
+        const arrays = ['widgets', 'roles', 'excludes', 'devices', 'values', 'siteCardMenuLinks'];
+        const renameTable = { permittedRoles: 'roles' };
+        // Reading references
+        xml.querySelectorAll('references > reference').forEach((tag) => {
+          config.references[tag.id] = applyDeserializedXMLTransforms(deserialize(tag.innerHTML), {
+            arrays,
+            renameTable
+          });
+        });
+        // Replacing references
+        xml.querySelectorAll('configuration > reference').forEach((tag) => {
+          tag.outerHTML = xml.querySelector(`references > reference[id='${tag.id}']`).innerHTML;
+        });
```

Comment:
This inner query is not the most efficient to do it on a loop. Let's flip this around: iterate through `:scope > references > reference` (you queried it before, just reuse same query result), and replace all of the `configuration > reference` that match the current reference id.
Also, try to cache the `config > reference` and query the selection instead of the whole document.